CIP-????/README.md

```diff
@@ -0,0 +1,294 @@
+
+However, in the interest of long-term API evolution we propose to add a flexible hashing interface:
+
+```
+data SignAlgo = SHA2_256
+              | SHA3_256
+              | Blake2
+
+hashWithAlgo :: HashAlgo -> BuiltinByteString -> BuiltinByteString
+```
+
+This will also allow for gradual _opt-in_ migration of scripts as the blockchain evolves.
+
+### Long term crypto API evolution
+
+It would be desirable to also offer a long-term API that would not need any changes
+as we migrate blockchain to new algorithms.
+
+This would be to provide an implicit embedding of a most secure available hashing algorithm
+within the hash itself, and use of interface:
+
+```
+data LTSignature = LongTermSignature {
+                     ltsHashAlgo :: HashAlgo
+                   , ltsSignAlgo :: SignAlgo
+                   , ltsSignature :: BuiltinByteString
+                   }
+
+verifyLongTermSignature :: BuiltinByteString                       -- ^ public key
+                        -> LongTermSignature                       -- ^ message or hash
+                        -> BuiltinByteString                       -- ^ signature
+                        -> Bool  -- is hash valid
+
+isLongTermSignatureSecure :: LongTermSignature -> Bool
+```
+
+Supporting evolution of hash algorithms is much simpler:
+
+```
+-- | Long term hash
+data LongTermHash = LongTermHash {
+                      hashAlgo :: HashAlgo
+                    , hash     :: BuiltingByteString
+                    }
+
+mostSecureHashAlgo :: HashAlgo
+
+mostSecureHash :: BuiltinByteString -> LongTermHash
+mostSecureHash = hashWithAlgo mostSecureHash
+```
+
+This API needs additional support from the off-chain applications in order to:
+
+* provide correct public key for the given algorithm
+* help users to transfer their euTxOs to new signature algorithms when old algorithms are obsoleted.
+
+Note that `LongTermHash` and `LongTermSignature` can be implemented as standard library functions.
+
+## Rationale
+
+We consider the implementation trustworthy after passing long comment period,
+considering that many competitors were eliminated by publishing weaknesses.
+
+An alternative approach could be to provide low-level primitives from the reference library,
+but that would decrease compatibility and security, since the reference implementation
+will likely be supported by other software.
+
+While lattice cryptography may require increased signature sizes, security-conscious
+users will be willing to pay the price.
+
+## Backward Compatibility
+
+At the Plutus Core level, implementing this proposal induces no
+backwards-incompatibility: the proposed new primitives do not break any existing
+functionality or affect any other builtins. Likewise, at levels above Plutus
+Core (such as `PlutusTx`), no existing functionality should be affected.
+
+On-chain, this requires a hard fork.
+
+## Path to Active
+
+An implementation will require the following steps:
+
+1. Linking to PQC library.
```

Comment:
Which library? That is a big deal!

Reply:
```suggestion
1. Linking to PQC library. We propose [PQClean](https://github.com/PQClean/PQClean) since it is a robust distribution of NIST-selected candidates. It also aims at cleanness and ease of formal verification.
```

Reply:
The reference C implementations of Post Quantum Crypto candidates from NIST are included in the library that facilitates their use.